Add RRef Python Helper to launch function on the referenced object

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36619 Add RRef Python Helper to launch function on the referenced object**

With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

An alternative design is to expose PyRRef as RRefBase and then
implement everything in a new Python RRef class. However, the RRef
class cannot directly inherit from PyRRef/RRefBase, otherwise we
will need to let pyRemote* C++ functions to load RRef from Python
and return an RRef instance. It is possible to let RRef hold a
instance of PyRRef instead of inherit from it, but this does not
look like a elegant design, as we will have RRef holding PyRRef and
PyRRef holding the C++ RRef. Another alternative is to use dynamic
method loading, by installing member methods to PyRRef instances.
However, this would require different solutions to handle
`RRef(data)` and `rpc.remote(...)`. Base on the above thinking, we
decided to go with the current implementation for simplicity and we
can also keep all RRef-related APIs in one place.

Differential Revision: [D21028333](https://our.internmc.facebook.com/intern/diff/D21028333)

Commit:
Add RRef Python Helper to launch function on the referenced object

With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

[ghstack-poisoned]

Commit:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

[ghstack-poisoned]

Commit:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

[ghstack-poisoned]

Commit:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

[ghstack-poisoned]

Commit:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

Differential Revision: [D21028333](https://our.internmc.facebook.com/intern/diff/D21028333)

[ghstack-poisoned]

zzzwen commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -214,6 +214,78 @@ PyObject* rpc_init(PyObject* /* unused */) {
+                      >>> from torch.distributed import rpc
+                      >>> rref = rpc.remote("worker1", torch.add, args=(torch.zeros(2, 2), 1))
+                      >>> rref.rpc_async().wait().size()  # returns torch.Size([2, 2])
+                      >>> rref.rpc_async().wait().view(1, 4)  # returns tensor([[1., 1., 1., 1.]])
```
should this be rref.rpc_async().view(1, 4).wait()

mrshenli replied:
Ahh, yes, thanks for the catch!!

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -214,6 +214,78 @@ PyObject* rpc_init(PyObject* /* unused */) {
+              py::call_guard<py::gil_scoped_release>(),
+              R"(
+                  Create a helper proxy to easily launch an ``rpc_sync`` using
+                  the owner of the RRef as the destination to run functions on
```
Wondering, is there a use case to eventually support these APIs on non-RRef owner nodes?

mrshenli replied:
good point! we could let this function take a `dst` arg to support that.

zzzwen replied:
If the plan is to implicitly pull the object to `dst` and run the method, we should be cautious 
 with the API. In pyper use case, we actually add extra assert to make sure code is always on its owner.

mrshenli replied:
I see. Then, let's not add that to this PR. We can add it later if we see real use cases and request for it.

xush6528 reviewed (changes requested):
.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -240,7 +312,8 @@ PyObject* rpc_init(PyObject* /* unused */) {
           // not releasing GIL to avoid context switch
-          .def("__str__", &PyRRef::str);
+          .def("__str__", &PyRRef::str)
+          .def("rpc_sync", &PyRRef::createRRefProxy);
```
ahh, yes, this is vestige for debugging, let me remove.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -214,6 +214,78 @@ PyObject* rpc_init(PyObject* /* unused */) {
                   If the current node is the owner, returns a reference to the
                   local value. Otherwise, throws an exception.
               )")
+          .def(
```
Yes, one alternative is making this `RRefBase` and then inherit it in Python and create a Python `RRef` class. But if we go with that approach, we would need to let `C++` `pyRemote*` load and return the Python version of `RRef`. This would also need some minor modification to serde.  

another alternative is instead of inheriting, we can let the Python `RRef` holds an instance of the C++ `RRefBase`. I am not sure if this a clean solution. 

The main reason for choosing the current implementation is for simplicity, trying to keep all Python RRef APIs in one place. Let me know if you prefer to change it to inheritance or composition.

Commit:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

An alternative design is to expose PyRRef as RRefBase and then
implement everything in a new Python RRef class. However, the RRef
class cannot directly inherit from PyRRef/RRefBase, otherwise we
will need to let pyRemote* C++ functions to load RRef from Python
and return an RRef instance. It is possible to let RRef hold a
instance of PyRRef instead of inherit from it, but this does not
look like a elegant design, as we will have RRef holding PyRRef and
PyRRef holding the C++ RRef. Another alternative is to use dynamic
method loading, by installing member methods to PyRRef instances.
However, this would require different solutions to handle
`RRef(data)` and `rpc.remote(...)`. Base on the above thinking, we
decided to go with the current implementation for simplicity and we
can also keep all RRef-related APIs in one place.

Differential Revision: [D21028333](https://our.internmc.facebook.com/intern/diff/D21028333)

[ghstack-poisoned]

mrshenli committed:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

An alternative design is to expose PyRRef as RRefBase and then
implement everything in a new Python RRef class. However, the RRef
class cannot directly inherit from PyRRef/RRefBase, otherwise we
will need to let pyRemote* C++ functions to load RRef from Python
and return an RRef instance. It is possible to let RRef hold a
instance of PyRRef instead of inherit from it, but this does not
look like a elegant design, as we will have RRef holding PyRRef and
PyRRef holding the C++ RRef. Another alternative is to use dynamic
method loading, by installing member methods to PyRRef instances.
However, this would require different solutions to handle
`RRef(data)` and `rpc.remote(...)`. Base on the above thinking, we
decided to go with the current implementation for simplicity and we
can also keep all RRef-related APIs in one place.

Differential Revision: [D21028333](https://our.internmc.facebook.com/intern/diff/D21028333)

[ghstack-poisoned]

mrshenli committed:
Update on "Add RRef Python Helper to launch function on the referenced object"


With this PR, applications no longer need to create dedicated helpers
to run functions on the object referenced by an RRef. Instead,
`rref.rpc_sync().some_func()` will use `rpc_sync` to run `some_func`
on the owner of the RRef using the object referenced by the RRef.
Similar helpers for `rref.rpc_async().some_func()` and
`rref.remote().some_func()` are also added.

An alternative design is to expose PyRRef as RRefBase and then
implement everything in a new Python RRef class. However, the RRef
class cannot directly inherit from PyRRef/RRefBase, otherwise we
will need to let pyRemote* C++ functions to load RRef from Python
and return an RRef instance. It is possible to let RRef hold a
instance of PyRRef instead of inherit from it, but this does not
look like a elegant design, as we will have RRef holding PyRRef and
PyRRef holding the C++ RRef. Another alternative is to use dynamic
method loading, by installing member methods to PyRRef instances.
However, this would require different solutions to handle
`RRef(data)` and `rpc.remote(...)`. Base on the above thinking, we
decided to go with the current implementation for simplicity and we
can also keep all RRef-related APIs in one place.

Differential Revision: [D21028333](https://our.internmc.facebook.com/intern/diff/D21028333)

[ghstack-poisoned]

stas00 commented:
Should this tutorial  https://pytorch.org/tutorials/intermediate/rpc_tutorial.html be updated to reflect reality of this change? It currently links to https://github.com/pytorch/pytorch/issues/31743 with:
> For example, it will be easier if applications can directly call RRef.some_func(*arg) which will then translate to RPC to the RRef owner. The progress on this API is tracked in pytorch/pytorch#31743.
